fix docker service stop

nicbus opened the pull request:
This PR changes the order in which services are stopped when the docker container is shut down.

With the previous order (based on `ls` output) I was getting the following logs:
```
shutting down container
4-2024-02-01T22:07:20Z tor: Thread interrupt
4-2024-02-01T22:07:20Z Shutdown: In progress...
4-2024-02-01T22:07:20Z RemoveLocal(g4xk3jxhqac4qqexdrnhk2mnd4c6dnmufvt4mifkwiyipvq7xzj5i3qd.onion:18444)
4-2024-02-01T22:07:20Z opencon thread exit
4-2024-02-01T22:07:20Z torcontrol thread exit
4-2024-02-01T22:07:20Z addcon thread exit
4-2024-02-01T22:07:20Z net thread exit
4-2024-02-01T22:07:20Z msghand thread exit
2-2024-02-01T22:07:20.430+00:00 - WARN - reconnecting to bitcoind: failed to read cookie from "/data/bitcoin/regtest/.cookie"
2-2024-02-01T22:07:23.431+00:00 - WARN - failed to connect daemon at 127.0.0.1:18443: Connection refused (os error 111)
2-2024-02-01T22:07:26.431+00:00 - WARN - failed to connect daemon at 127.0.0.1:18443: Connection refused (os error 111)
kill: run: bitcoin: (pid 75) 35s, want down, got TERM
2-2024-02-01T22:07:27.660+00:00 - TRACE - stop accepting new RPCs
2-2024-02-01T22:07:27.660+00:00 - TRACE - closing 0 RPC connections
2-2024-02-01T22:07:27.660+00:00 - TRACE - RPC connections are closed
2-2024-02-01T22:07:27.660+00:00 - TRACE - RPC server is stopped
2-2024-02-01T22:07:27.660+00:00 - ERROR - server failed: Error: Iterrupted by signal 15
2-   0: <unknown>
2-   1: <unknown>
2-   2: <unknown>
2-   3: <unknown>
2-   4: <unknown>
2-   5: <unknown>
2-   6: <unknown>
2-   7: <unknown>
2-   8: <unknown>
2-   9: <unknown>
2-  10: __libc_start_main
2-  11: <unknown>
2-
2-
2-2024-02-01T22:07:27.662+00:00 - TRACE - deregistering event source from poller
ok: down: electrs: 1s, normally up
ok: down: nginx: 0s, normally up
ok: down: prerenderer: 0s, normally up
ok: down: socat: 1s, normally up
ok: down: tor: 0s, normally up
```
Also, when starting the container again, bitcoind was always resuming from height 0, instead of the one it was at before the stop was requested.

With the change in this PR I am getting:
```
shutting down container
ok: down: nginx: 0s, normally up
6-Feb 01 22:11:19.000 [notice] Catching signal TERM, exiting cleanly.
4-2024-02-01T22:11:19Z RemoveLocal(3z4bdbvidw6aq46z7gxkj6nrvfukatdw4bvlyh2lakz5ranswsga3jad.onion:18444)
ok: down: tor: 1s, normally up
ok: down: socat: 0s, normally up
ok: down: websocket: 0s, normally up
ok: down: prerenderer: 1s, normally up
ok: down: electrs: 0s, normally up
ok: down: bitcoin: 1s, normally up
```
Also, when starting the container again, bitcoind resumes from the height it was before the stop was requested.

Notes:
- tested using `bitcoin-regtest`
- I also took the liberty to address some (shellcheck) lint warnings

shesek commented:
Thanks for the PR!

If I understand correctly, the necessary change is to have `electrs` shut down prior to `bitcoin`, right?

Perhaps it would be easier to explicitly issue a `force-stop` to `electrs` first, then iterate over the services in the directory? That way there's no need to maintain an updated list of services in the script and no need to check for a diff.

nicbus commented:
Only stopping the required services is better indeed. It's not just `electrs` though, so I made more tests and isolated `socat` as well (which needs to be stopped first).

Updated and rebased.

1alexbc1 reviewed (commented):
Bca

philippem commented on contrib/runit_boot.sh:
```diff
@@ -3,14 +3,20 @@
+  # first shut down socat (avoid "Iterrupted by signal 15" error)
+  sv -w "$SVWAIT" force-stop socat
+
+  # next shut down electrs (avoid long wait + losing blocks in bitcoind)
```
this is an improvement to shutdown/restart robustness.

philippem commented on contrib/runit_boot.sh:
```diff
@@ -38,8 +44,9 @@ echo "Started runsvdir, PID is $RUNSVDIR"
 sleep 5
-for _srv in $(ls -1 /etc/service); do
-    sv status $_srv
+mapfile -t etc_services < <(ls -1 /etc/service)
```
this mapfile executes once when the bash script begins (at container launch). The array is initialized at that time.  This subtly changes the behaviour of "shutdown"

nicbus replied:
I was not thinking about people changing services inside a running container so I didn't notice this subtlety, nice catch, thanks!

philippem commented:
this is a good change, can you rebase and remove the use of the bash array?

nicbus commented:
thanks for the review, rebased and restored the original way to list services

shesek commented:
The changes look good :+1:

I'm wondering, though, perhaps we should give bitcoind even more time to shutdown? Say, 5 minutes?

Shutting down cleanly can take some time if there's a lots of data left to flush.

nicbus commented:
> I'm wondering, though, perhaps we should give bitcoind even more time to shutdown? Say, 5 minutes?

added a commit to do that and rebased on master

shesek commented:
Tested ACK 3de4458e04fdf2738ee46c7f8fda335c16bb5cd5. Merged.

Thank you for the contribution @nicbus!
